addon-test-support/setup-middleware-reporter.ts

```diff
@@ -1,36 +1,70 @@
 import QUnit from 'qunit';
-import { getContext, getTestMetadata } from '@ember/test-helpers';
-import { AxeResults } from 'axe-core';
+import {
+  currentRouteName,
+  currentURL,
+  getContext,
+  getTestMetadata,
+} from '@ember/test-helpers';
+import { AxeResults, Result } from 'axe-core';
 import { setCustomReporter } from './reporter';
 import { DEBUG } from '@glimmer/env';
 
 interface AxeTestResult {
   moduleName: string;
   testName: string;
-  helperName: string;
+  urls: string[];
+  routes: string[];
+  helpers: string[];
   stack: string;
-  axeResults: AxeResults;
+  violations: Result[];
 }
 
 export const TEST_SUITE_RESULTS: AxeTestResult[] = [];
 
-export function buildResult(axeResults: AxeResults) {
-  let { module, testName } = QUnit.config.current;
-  let testMetaData = getTestMetadata(getContext());
+let currentTestResult: AxeTestResult | undefined = undefined;
+let currentViolationsMap = new Map<string, Result>();
 
-  let stack = (!DEBUG && new Error().stack) || '';
+export async function middlewareReporter(axeResults: AxeResults) {
+  if (axeResults.violations.length === 0) {
+    return;
+  }
+
+  if (!currentTestResult) {
+    let { module, testName } = QUnit.config.current;
+    let testMetaData = getTestMetadata(getContext());
+
+    let stack = (!DEBUG && new Error().stack) || '';
+
+    currentTestResult = {
+      moduleName: module.name,
+      testName,
+      urls: [currentURL()],
+      routes: [currentRouteName()],
+      helpers: testMetaData.usedHelpers,
+      stack,
+      violations: [],
+    };
+  }
+
+  axeResults.violations.forEach((violation) => {
+    let rule = currentViolationsMap.get(violation.id);
 
-  return {
-    moduleName: module.name,
-    testName,
-    helperName: testMetaData.usedHelpers.pop() || '',
-    stack,
-    axeResults,
-  };
+    if (rule === undefined) {
+      currentViolationsMap.set(violation.id, violation);
+    } else {
+      rule.nodes.push(...violation.nodes);
+    }
+  });
 }
 
-export async function middlewareReporter(axeResults: AxeResults) {
-  TEST_SUITE_RESULTS.push(buildResult(axeResults));
+export function pushTestResult() {
+  if (typeof currentTestResult !== 'undefined') {
+    currentTestResult.violations = [...currentViolationsMap.values()];
+    TEST_SUITE_RESULTS.push(currentTestResult);
+
+    currentTestResult = undefined;
+    currentViolationsMap = new Map();
```

Comment:
It's vaguely odd to reset one to `new Map()` and the other to `undefined`. I guess I'd prefer to set both back to `undefined` here, then initialize inside the `if (!currentTestResult) {` branch above.

Reply:
🤷

Reply:
I changed it to make it more consistent.